[BUG] While instrumenting the rclone repo I got a SIGSEGV

<!--
If you have identified a specific bug in the orchestrion codebase, please describe it here. If the
issue requires debugging or investigation of your specific private setup, that is best accomplished via support. Please contact Datadog [support](http://docs.datadoghq.com/help/).
-->

**Version of orchestrion**
master

<!-- If you aren't using the [latest version](https://github.com/DataDog/orchestrion/releases) of orchestrion, try upgrading first to see if your issue has already been resolved -->

**Describe what happened:**
I'm working on the CI Visibility instrumentation and while testing with the `rclone` repo I hit the following panic in orchestrion:

```go
...
ok  	github.com/rclone/rclone/lib/kv	0.775s
ok  	github.com/rclone/rclone/lib/mmap	0.992s
ok  	github.com/rclone/rclone/lib/pacer	2.020s
# github.com/rclone/rclone/librclone
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x8 pc=0x1015e059c]

goroutine 1 [running]:
github.com/datadog/orchestrion/internal/injector.filterExtraneousImports(...)
	../../tonyredondo/orchestrion/internal/injector/imports.go:56
github.com/datadog/orchestrion/internal/injector.canonicalizeImports(0x14000276a90)
	../../tonyredondo/orchestrion/internal/injector/imports.go:25 +0x21c
github.com/datadog/orchestrion/internal/injector.(*Injector).writeModifiedFile(0x1400017ea80, 0x140001df7a0, 0x14000276a90)
	../../tonyredondo/orchestrion/internal/injector/write.go:23 +0x34
github.com/datadog/orchestrion/internal/injector.(*Injector).injectFile(0x1400017ea80, 0x140001df7a0, 0x14000276a90)
	../../tonyredondo/orchestrion/internal/injector/injector.go:132 +0xa4
github.com/datadog/orchestrion/internal/injector.(*Injector).InjectFiles(0x1400017ea80, {0x1400031cdc0, 0x3, 0x15})
	../../tonyredondo/orchestrion/internal/injector/injector.go:98 +0x33c
github.com/datadog/orchestrion/internal/toolexec/aspect.Weaver.OnCompile({{0x1400003a214?, 0x100cec920?}}, 0x1400030a300)
	../../tonyredondo/orchestrion/internal/toolexec/aspect/oncompile.go:82 +0x684
github.com/datadog/orchestrion/internal/toolexec/proxy.ProcessCommand[...]({0x1019b3b18?, 0x1400030a300?}, 0x14000229708?)
	../../tonyredondo/orchestrion/internal/toolexec/proxy/proxy.go:74 +0x44
github.com/datadog/orchestrion/internal/cmd.init.func4(0x14000229978?)
	../../tonyredondo/orchestrion/internal/cmd/toolexec.go:56 +0x2ac
github.com/urfave/cli/v2.(*Command).Run(0x102004c20, 0x140001314c0, {0x1400031c580, 0x16, 0x16})
	../../../../go/pkg/mod/github.com/urfave/cli/v2@v2.27.4/command.go:276 +0x600
github.com/urfave/cli/v2.(*Command).Run(0x1400031c160, 0x14000131280, {0x140001c2000, 0x17, 0x18})
	../../../../go/pkg/mod/github.com/urfave/cli/v2@v2.27.4/command.go:269 +0x83c
github.com/urfave/cli/v2.(*App).RunContext(0x1400017c600, {0x1019b0e40, 0x1020c5c60}, {0x140001c2000, 0x17, 0x18})
	../../../../go/pkg/mod/github.com/urfave/cli/v2@v2.27.4/app.go:333 +0x51c
github.com/urfave/cli/v2.(*App).Run(...)
	../../../../go/pkg/mod/github.com/urfave/cli/v2@v2.27.4/app.go:307
main.main()
	../../tonyredondo/orchestrion/main.go:162 +0x71c
ok  	github.com/rclone/rclone/lib/pool	2.385s
ok  	github.com/rclone/rclone/lib/random	0.745s
ok  	github.com/rclone/rclone/lib/ranges	0.885s
...
```

**Describe what you expected:**
not to panic

**Steps to reproduce the issue:**
<!-- If you can include a minimal code sample to reproduce the issue it can significantly reduce resolution time -->
Just ran `/Users/tony.redondo/repos/github/tonyredondo/orchestrion/orchestrion go test ./...` 

**Additional environment details (Version of Go, Operating System, etc.):**

```bash
❯ go version
go version go1.23.0 darwin/arm64
```

```bash
❯ sw_vers
ProductName:		macOS
ProductVersion:		14.6.1
BuildVersion:		23G93
```

--- a/internal/injector/imports.go
+++ b/internal/injector/imports.go
@@ -9,6 +9,7 @@ import (
 	"go/token"
 	"strconv"
 
+	"github.com/DataDog/orchestrion/internal/log"
 	"github.com/dave/dst"
 	"github.com/dave/dst/dstutil"
 )
@@ -20,7 +21,7 @@ import (
 // To do so, it modifies the AST file so that it only includes a single import per path, using the
 // first non-empty alias found.
 func canonicalizeImports(file *dst.File) {
-	specsByPath := importSpecsByImportPath(file)
+	specsByPath := importSpecsByImportPath(file.Imports)
 
 	retain := filterExtraneousImports(specsByPath)
 
@@ -32,16 +33,16 @@ func canonicalizeImports(file *dst.File) {
 	filterDecls(file, retain)
 }
 
-func importSpecsByImportPath(file *dst.File) map[string][]*dst.ImportSpec {
-	byPath := make(map[string][]*dst.ImportSpec, len(file.Imports))
+func importSpecsByImportPath(imports []*dst.ImportSpec) map[string][]*dst.ImportSpec {
+	byPath := make(map[string][]*dst.ImportSpec, len(imports))
 
-	for _, imp := range file.Imports {
+	for _, imp := range imports {
 		path, err := strconv.Unquote(imp.Path.Value)
 		if err != nil {
+			log.Debugf("failed to unquote import path %q: %v\n", imp.Path.Value, err)
 			continue
 		}
-		list := append(byPath[path], imp)
-		byPath[path] = list
+		byPath[path] = append(byPath[path], imp)
 	}
 
 	return byPath
@@ -53,10 +54,22 @@ func filterExtraneousImports(byPath map[string][]*dst.ImportSpec) map[*dst.Impor
 	for _, specs := range byPath {
 		retain := specs[0]
 		for _, spec := range specs[1:] {
-			if (spec.Name == nil && (retain.Name == nil || retain.Name.Name != "_")) || spec.Name.Name == "_" {
+			// Preference order is: spec.Name == nil < spec.Name.Name == "_" < spec.Name.Name != "_"
+			if spec.Name == nil {
 				continue
 			}
-			retain = spec
+
+			if retain.Name == nil && spec.Name != nil {
+				retain = spec
+				continue
+			}
+
+			if retain.Name.Name == "_" && spec.Name.Name != "_" {
+				retain = spec
+				continue
+			}
+
+			// We found a non-empty alias, no need to look further.
 			break
 		}
 		result[retain] = struct{}{}

--- a/internal/injector/imports_test.go
+++ b/internal/injector/imports_test.go
@@ -1,0 +1,131 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+package injector
+
+import (
+	"testing"
+
+	"github.com/dave/dst"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFilterExtraneousImports(t *testing.T) {
+	qualifedImport := func(name string) *dst.ImportSpec {
+		return &dst.ImportSpec{Name: &dst.Ident{Name: name}}
+	}
+
+	for _, tc := range []struct {
+		name     string
+		in       []*dst.ImportSpec
+		expected *dst.ImportSpec
+	}{
+		{
+			name: "simple",
+			in: []*dst.ImportSpec{
+				qualifedImport("test"),
+			},
+			expected: qualifedImport("test"),
+		},
+		{
+			name: "multiple-qualified",
+			in: []*dst.ImportSpec{
+				qualifedImport("test1"),
+				qualifedImport("test2"),
+			},
+			expected: qualifedImport("test1"),
+		},
+		{
+			name: "one-qualified-first",
+			in: []*dst.ImportSpec{
+				qualifedImport("test1"),
+				{Name: &dst.Ident{Name: "_"}},
+			},
+			expected: qualifedImport("test1"),
+		},
+		{
+			name: "one-qualified-last",
+			in: []*dst.ImportSpec{
+				qualifedImport("test1"),
+			},
+			expected: qualifedImport("test1"),
+		},
+		{
+			name: "one-qualified-first-with-nil",
+			in: []*dst.ImportSpec{
+				qualifedImport("test1"),
+				{Name: nil},
+			},
+			expected: qualifedImport("test1"),
+		},
+		{
+			name: "one-qualified-last-with-nil",
+			in: []*dst.ImportSpec{
+				{Name: nil},
+				qualifedImport("test1"),
+			},
+			expected: qualifedImport("test1"),
+		},
+		{
+			name: "complex-1",
+			in: []*dst.ImportSpec{
+				{Name: nil},
+				{Name: &dst.Ident{Name: "_"}},
+				qualifedImport("test1"),
+				{Name: nil},
+			},
+			expected: qualifedImport("test1"),
+		},
+		{
+			name: "complex-2",
+			in: []*dst.ImportSpec{
+				{Name: &dst.Ident{Name: "_"}},
+				{Name: nil},
+				{Name: nil},
+			},
+			expected: &dst.ImportSpec{Name: &dst.Ident{Name: "_"}},
+		},
+		{
+			name: "complex-3",
+			in: []*dst.ImportSpec{
+				{Name: &dst.Ident{Name: "_"}},
+				{Name: nil},
+				{Name: nil},
+			},
+			expected: &dst.ImportSpec{Name: &dst.Ident{Name: "_"}},
+		},
+		{
+			name: "complex-4",
+			in: []*dst.ImportSpec{
+				{Name: nil},
+				{Name: &dst.Ident{Name: "_"}},
+				{Name: nil},
+			},
+			expected: &dst.ImportSpec{Name: &dst.Ident{Name: "_"}},
+		},
+		{
+			name: "complex-5",
+			in: []*dst.ImportSpec{
+				{Name: nil},
+				qualifedImport("test1"),
+				{Name: &dst.Ident{Name: "_"}},
+				{Name: nil},
+				qualifedImport("test2"),
+			},
+			expected: qualifedImport("test1"),
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			out := filterExtraneousImports(map[string][]*dst.ImportSpec{
+				"test": tc.in,
+			})
+
+			require.Len(t, out, 1)
+			for v := range out {
+				require.EqualValues(t, tc.expected, v)
+			}
+		})
+	}
+}

--- a/internal/injector/imports_test.go
+++ b/internal/injector/imports_test.go
@@ -1,7 +1,7 @@
 // Unless explicitly stated otherwise all files in this repository are licensed
 // under the Apache License Version 2.0.
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
-// Copyright 2024 Datadog, Inc.
+// Copyright 2023-present Datadog, Inc.
 
 package injector
 

--- a/internal/injector/injector.go
+++ b/internal/injector/injector.go
@@ -96,7 +96,7 @@ func (i *Injector) InjectFiles(files []string) (map[string]InjectedFile, error) 
 
 	wg.Add(len(astFiles))
 	for idx, astFile := range astFiles {
-		go func(astFile *ast.File) {
+		go func(idx int, astFile *ast.File) {
 			defer wg.Done()
 
 			decorator := decorator.NewDecoratorWithImports(fset, i.ImportPath, gotypes.New(uses))
@@ -123,7 +123,7 @@ func (i *Injector) InjectFiles(files []string) (map[string]InjectedFile, error) 
 			resultMu.Lock()
 			defer resultMu.Unlock()
 			result[files[idx]] = res.InjectedFile
-		}(astFile)
+		}(idx, astFile)
 	}
 	wg.Wait()
 

--- a/internal/injector/write.go
+++ b/internal/injector/write.go
@@ -21,10 +21,10 @@ import (
 // writeModifiedFile writes the modified file to disk after having restored it to Go source code,
 // and returns the path to the modified file.
 func (i *Injector) writeModifiedFile(decorator *decorator.Decorator, file *dst.File) (string, error) {
-	log.Tracef("writing modified file %q\n", decorator.Filenames[file])
-	canonicalizeImports(file)
-
 	filename := decorator.Filenames[file]
+
+	log.Tracef("writing modified file %q\n", filename)
+	canonicalizeImports(file)
 
 	if err := lineinfo.AnnotateMovedNodes(decorator, file, i.newRestorer); err != nil {
 		return filename, fmt.Errorf("annotating moved nodes in %q: %w", filename, err)

--- a/internal/injector/write.go
+++ b/internal/injector/write.go
@@ -13,6 +13,7 @@ import (
 	"path/filepath"
 
 	"github.com/DataDog/orchestrion/internal/injector/lineinfo"
+	"github.com/DataDog/orchestrion/internal/log"
 	"github.com/dave/dst"
 	"github.com/dave/dst/decorator"
 )
@@ -20,6 +21,7 @@ import (
 // writeModifiedFile writes the modified file to disk after having restored it to Go source code,
 // and returns the path to the modified file.
 func (i *Injector) writeModifiedFile(decorator *decorator.Decorator, file *dst.File) (string, error) {
+	log.Tracef("writing modified file %q\n", decorator.Filenames[file])
 	canonicalizeImports(file)
 
 	filename := decorator.Filenames[file]

--- a/tools/headercheck/header_check.go
+++ b/tools/headercheck/header_check.go
@@ -21,9 +21,6 @@ func main() {
 		"// This product includes software developed at Datadog (https://www.datadoghq.com/).",
 		"// Copyright 2023-present Datadog, Inc.",
 	}
-	skip := map[string]struct{}{
-		"event_string.go": {}, // Auto generated file
-	}
 	pwd, err := os.Getwd()
 	if err != nil {
 		log.Fatal(err)
@@ -35,9 +32,6 @@ func main() {
 			log.Fatal(err)
 		}
 		if !strings.HasSuffix(info.Name(), ".go") {
-			return nil
-		}
-		if _, skipped := skip[info.Name()]; skipped {
 			return nil
 		}
 		file, err := os.Open(path)

